tools/import.bzl

```diff
@@ -0,0 +1,84 @@
+# Copyright 2018 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+def _get_basename(url):
+  return url[url.rindex("/") + 1:]
+
+def _bazel_external_dependency_archive_impl(repository_ctx):
+  build_lines = [
+      "filegroup(",
+      "    name = \"%s\"," % repository_ctx.name,
+      "    srcs = [",
+  ]
+
+  for sha256 in repository_ctx.attr.srcs:
+    urls = repository_ctx.attr.srcs[sha256]
+    basename = sha256
+    repository_ctx.download(
+        url = urls,
+        sha256 = sha256,
+        output = basename,
+    )
+    build_lines.append("        \"%s\",  # %s" % (basename, urls[0]))
+
+  build_lines += [
+      "    ],",
+      "    visibility = [\"//visibility:public\"],",
+      ")",
+  ]
+
+  repository_ctx.file("BUILD", "\n".join(build_lines))
+
+bazel_external_dependency_archive = repository_rule(
+    attrs = {
+        "srcs": attr.string_list_dict(allow_empty = False),
```

Comment:
Why is this a dict? AFAICT if I add two entries they will both be part of the same external repository in the scratch workspace right? Am I reading it wrong?

Reply:
It's a dictionary of sha256 digests to download URLs.  The download cache works with digests, so the download URLs themselves are irrelevant in the scratch workspace, only the sha256 digests matter.

To clarify things a bit, the cache is only a directory with the following descendants:
content_addressable/sha256/<sha256 digest>/file

So what I do is I take the downloaded file for each external dependency, which I name as its sha256 digest, put it as a runfile and construct by hand a repository cache with the layout above.

This requires of course that the corresponding repository rule uses a sha256 digest that is explicitly given as a rule attribute.  That's why, in Java, maven_jar won't work, but java_import_external will.  (maven_jar should not be used anymore, IMO, because among other things the sha-1 checksum is optional and it is not possible to specify transitive dependencies, and I think it should be deprecated.)

I suggest you look at the two new test cases in WorkspaceDriverIntegrationTest to get a better perspective on all that.

Reply:
I looked at all of the code but I think that what I was missing (and maybe should be added as a comment) is the layout of the cache. I’m good with respect to this comment.  I think it only strengthens my other comment about having a test (maybe the existing ones) which has two entries in the dict